---
title: Generator.prototype.throw()
slug: Web/JavaScript/Reference/Global_Objects/Generator/throw
tags:
  - ECMAScript 2015
  - Generator
  - JavaScript
  - Довідка
  - метод
translation_of: Web/JavaScript/Reference/Global_Objects/Generator/throw
---
<div>{{JSRef}}</div>

<p>Метод <code><strong>throw()</strong></code> відновлює виконання генератора викиданням у ньому помилки та повертає об'єкт з двома властивостями, <code>done</code> та <code>value</code>.</p>

<h2 id="Синтаксис">Синтаксис</h2>

<pre class="syntaxbox"><code><var>gen</var>.throw(exception)</code></pre>

<h3 id="Параметри">Параметри</h3>

<dl>
 <dt><code>exception</code></dt>
 <dd>Виняток, який треба викинути. З метою налагоджування корисно робити його <code>instanceof</code> {{jsxref("Error")}}.</dd>
</dl>

<h3 id="Значення_що_повертається">Значення, що повертається</h3>

<p><code><a href="/uk/docs/Web/JavaScript/Reference/Global_Objects/Object">Об'єкт</a></code> з двома властивостями:</p>

<ul>
 <li><code>done</code> (булеве значення)

  <ul>
   <li>Має значення <code>true</code>, якщо ітератор досяг кінця послідовності перебору. В цьому випадку <code>value</code> за бажанням визначає <em>повернене значення</em> ітератора.</li>
   <li>Має значення <code>false</code>, якщо ітератор зміг видати наступне значення послідовності. Це аналогічно тому, щоб взагалі не вказувати значення властивості <code>done</code>.</li>
  </ul>
 </li>
 <li>будь-яке значення JavaScript, повернене ітератором. Може бути пропущене, коли <code>done</code> дорівнює <code>true</code>.</li>
</ul>

<h2 id="Приклади">Приклади</h2>

<h3 id="Використання_throw">Використання <code>throw()</code></h3>

<p>Наступний приклад демонструє простий генератор та помилку, що викидається методом <code>throw</code>. Помилку, як завжди, можна перехопити блоком <code><a href="/uk/docs/Web/JavaScript/Reference/Statements/try...catch">try...catch</a></code>.</p>

<pre class="brush: js">function* gen() {
  while(true) {
    try {
       yield 42;
    } catch(e) {
      console.log('Перехоплено помилку!');
    }
  }
}

var g = gen();
g.next();
// { value: 42, done: false }
g.throw(new Error('Щось пішло не за планом'));
// "Перехоплено помилку!"
// { value: 42, done: false }
</pre>

<h2 id="Специфікації">Специфікації</h2>

<table class="standard-table">
 <tbody>
  <tr>
   <th scope="col">Специфікація</th>
   <th scope="col">Статус</th>
   <th scope="col">Коментар</th>
  </tr>
  <tr>
   <td>{{SpecName('ES2015', '#sec-generator.prototype.throw', 'Generator.prototype.throw')}}</td>
   <td>{{Spec2('ES2015')}}</td>
   <td>Початкове визначення.</td>
  </tr>
  <tr>
   <td>{{SpecName('ESDraft', '#sec-generator.prototype.throw', 'Generator.prototype.throw')}}</td>
   <td>{{Spec2('ESDraft')}}</td>
   <td></td>
  </tr>
 </tbody>
</table>

<h2 id="Сумісність_з_веб-переглядачами">Сумісність з веб-переглядачами</h2>

<div>


<p>{{Compat("javascript.builtins.Generator.throw")}}</p>
</div>

<h2 id="Див._також">Див. також</h2>

<ul>
 <li><code><a href="/uk/docs/Web/JavaScript/Reference/Statements/function*">function*</a></code></li>
</ul>
